### PR TITLE
feat(ledger): enable VRF/KES validation in the block-processing pipeline (#1894)

### LINKED ISSUE #1894: Block processing pipeline: parallel decode and validate
Sync speed is ~3,905 slots/sec on preview, target is >10,000. Parallelizing block decode and validation is the main lever. The gouroboros pipeline/ package provides the foundation.

Phases:
1. Pipeline plumbing -- create pipeline in LedgerState, submit blocks, apply via chain.AddBlock. 2-worker decode, no validation yet.
2. Wire apply stage -- chain add, error handling, event publishing, sync progress logging.
3. Enable VRF/KES validation -- Eta0Provider for epoch nonce, validate workers.
4. Metrics integration -- pipeline stats counters and gauges.
5. Rollback coordination -- drain in-flight blocks before rollback.

Acceptance:
- 2x+ bulk sync throughput
- Tests pass with pipeline enabled and disabled
- Chain tip matches both modes after identical sync
- No goroutine leaks
- Rollback behavior identical with and without pipeline

Roadmap: Q3 2026

### PR DESCRIPTION
**Pull request by cryptodj413**
Closes #1894

Enables VRF/KES validation in the block-processing pipeline (gouroboros
parallel-validate pipeline) and works through the correctness issues that
surfaced integrating it with dingo's ledger read-chain path. Landed in 6
phases:

- **Phase 1** (c6142feb) — `feat(ledger)`: enable VRF/KES validation in the block-processing pipeline
- **Phase 2** (43fcacc0) — `fix(ledger)`: skip redundant admission-time VRF/KES re-check when the block pipeline validates
- **Phase 3** (3dc63793) — `docs(ledger)`: document why pipeline apply-stage wiring is deferred to #3227
- **Phase 4** (8b1ef5a4) — `fix(ledger)`: restart blockfetch when fork-resolution header-queue overflow leaves it undrained
- **Phase 5** (4c82d05a) — `feat(ledger)`: drain block-processing pipeline before rollback
- **Phase 6** (fcc32370) — `fix(ledger)`: drain block-processing pipeline error channel to prevent Byron-sync deadlock

## Start review here: Phase 6

Phase 6 was added after two rounds of review and is the part most worth a
close look. The first review round caught a critical, silent, permanent
deadlock: with `BlockPipelineValidateEnabled` on, gouroboros' pipeline pushes
every validate-stage error — including the always-expected Byron no-nonce
error — onto a bounded, undrained `errorsChan`. Nothing in dingo drained
that channel, so on any real chain sync (including the Byron era) it would
fill after ~1000 Byron blocks, every validate worker would then block
forever, and the ledger read-chain goroutine would hang permanently with no
error, log, or timeout.

Phase 6 fixes this with:
- a permanent drain goroutine, started right after the pipeline starts and
  joined cleanly on `Close()`
- two new Prometheus counters classifying expected-Byron vs.
  genuinely-unexpected errors
- a regression test pushing 4000 validation-failing blocks through the real
  pipeline to confirm it no longer hangs

This was re-reviewed and confirmed correct, race-clean, and non-regressive.
One minor non-blocking gap was noted as a fast-follow (not fixed here, does
not reintroduce the deadlock and does not affect correctness):
`blockPipelineEta0Provider` currently classifies a few rare, unrelated
`headerVerificationEpoch` failure branches under the same "expected Byron"
bucket as the true no-nonce case, slightly diluting the unexpected-error
signal for those edge cases.

## Testing

- `go test ./ledger/... -race` passes, including the new phase 6 regression
  test (4000 validation-failing blocks through the real pipeline)
- `golangci-lint run ./...`, `nilaway ./...`, `modernize ./...` clean
- `make import-boundaries` clean
- `ARCHITECTURE.md` updated for the pipeline drain/rollback behavior changes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Enables parallel VRF/KES validation in the block-processing pipeline and hardens replay/rollback to prevent deadlocks and stalls. Previously only decode ran in parallel and Byron-era validation errors could fill an undrained channel and hang; now validation runs in-pipeline, redundant admission-time crypto checks are skipped, errors are drained and classified, and rollback/fork paths coordinate with the pipeline.

- Adds `blockPipelineValidateEnabled` (CLI `--block-pipeline-validate-enabled`, env `DINGO_BLOCK_PIPELINE_VALIDATE_ENABLED`), off by default and requiring `blockPipelineEnabled`. Uses `gouroboros/pipeline` validate workers with an Eta0 provider.
- Skips admission-time stateless VRF/KES re-checks when the validate stage is on; blocks are still verified once before apply. State checks and epoch-cache advance still run.
- Continuously drains the pipeline’s error channel; classifies expected errors as eta0-unavailable (no epoch nonce) and deferred epoch-cache; exposes counters for expected vs unexpected errors; drains remaining results after batch failures.
- Drains the pipeline before rollback with a bounded timeout and adds a gather mutex to close a submit-after-rollback race.
- Restarts blockfetch when fork-resolution overflows the header queue; on restart failure, clears the queue and resyncs; fixes stale peer selection on retarget to avoid long stalls.
- Exposes pipeline gauges (decoded, validated, decode/validate errors, queue depth); apply-stage wiring remains deferred (sequential apply unchanged).

- Rollout
  - No default behavior change. To enable validation, set `blockPipelineEnabled: true` and `blockPipelineValidateEnabled: true`.
  - Monitor the new counters; eta0-unavailable and deferred epoch-cache errors are expected when syncing Byron or while the epoch cache catches up.

<sup>Written for commit fd51dff164684b357f28e34831a36ff463364aba. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/blinklabs-io/dingo/pull/3232?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added optional parallel VRF/KES validation during block-pipeline processing.
  * Added configuration support through YAML, environment variables, and CLI flags.
  * Added metrics for pipeline activity, queue depth, and validation errors.

* **Bug Fixes**
  * Improved recovery when fork queues overflow.
  * Prevented validation errors from blocking pipeline progress.
  * Coordinated pipeline shutdown before rollbacks to preserve state consistency.

* **Documentation**
  * Documented pipeline lifecycle, validation, retries, metrics, rollback handling, and operational limitations.

### CODE REVIEW

File: ledger/state.go [hunk 1]
```diff
@@ -3442,7 +3771,32 @@ func (ls *LedgerState) decodeReadChainBatch(
...
+			}
+			if !item.IsValid() {
+				ls.config.Logger.Error(
+					"block failed pipeline VRF/KES validation",
+					"slot", block.SlotNumber(),
+				)
+				return nil, false
+			}
+		}
+		decoded = append(decoded, block)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/state.go:3939]: Addressed in 97cf2edf: after a batch decode/validation failure, the remaining in-flight pipeline results are now drained before returning, so the pipeline goroutine cannot block on a full channel and the next block admission is not left stalled.

File: ledger/state.go [hunk 2]
```diff
@@ -2653,6 +2797,96 @@ func (ls *LedgerState) rollback(point ocommon.Point) error {
...
+	if ls.blockPipeline == nil {
+		return
+	}
+	waitCtx, cancel := context.WithTimeout(
+		ctx,
+		BlockPipelineRollbackDrainTimeout,
+	)
+	defer cancel()
+	start := time.Now()
+	if err := ls.blockPipeline.WaitForDrain(waitCtx); err != nil {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/state.go:2888]: Addressed in 97cf2edf: added blockPipelineGatherMutex serializing the reader's gather-then-submit sequence against Rollback, so a chainsync rollback can no longer interleave with an in-flight batch gather/submit and race the pipeline's internal state.

File: ledger/verify_header.go [hunk 3]
```diff
@@ -62,6 +62,19 @@ var (
...
+	// slot always means no Praos epoch nonce is available for it, which in
+	// practice only ever happens for Byron-era slots (see
+	// blockPipelineEta0Provider's doc comment) -- Byron has no Praos nonce
+	// at all, and the pipeline only validates blocks already committed to
+	// ls.chain, so a post-Byron slot's epoch/nonce data should already be
+	// present. drainBlockPipelineErrors uses this sentinel to log/count
+	// these as the expected case rather than at error level.
+	errBlockPipelineEta0Unavailable = errors.New(
+		"block-processing pipeline: epoch nonce unavailable",
+	)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_header.go]: Addressed in 97cf2edf: headerVerificationEpoch/blockPipelineEta0Provider now narrows the eta0-lookup-failure short-circuit to the genuine Byron no-nonce case only; other epoch/forecast/hard-fork-summary lookup errors are no longer swallowed and propagate as real validation errors instead of being silently treated as "skip, no nonce available".

File: ledger/verify_header.go [hunk 4]
```diff
@@ -1327,6 +1340,36 @@ func (ls *LedgerState) epochNonceHex(epochId uint64, nonce []byte) string {
...
+// cover them. Not advancing it also avoids mutating shared epoch-cache state
+// concurrently from multiple validate-stage worker goroutines, which all
+// call this function without any external synchronization between them.
+//
+// Byron-era slots have no Praos epoch nonce and always fail this lookup
+// (headerVerificationEpoch rejects an epoch with an empty nonce); callers
+// must treat a validation failure on a Byron-era block as expected, exactly
+// as the serial path's verifyBlockHeaderHex explicitly skips Byron blocks
+// rather than trying to validate them.
+func (ls *LedgerState) blockPipelineEta0Provider(slot uint64) (string, error) {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/verify_header.go:1382]: Same fix as the coderabbitai eta0-error-classification comment above, addressed in 97cf2edf -- see reply there for detail.

File: ledger/chainsync.go [hunk 5]
```diff
@@ -3114,6 +3186,53 @@ func (ls *LedgerState) restartQueuedBlockfetchAfterForkLocked(
...
+		return
+	}
+	if err := ls.startQueuedBlockfetchLocked(connId, pending); err != nil {
+		ls.config.Logger.Warn(
+			"failed to start blockfetch after fork-resolution queue overflow",
+			"component", "ledger",
+			"error", err,
+			"connection_id", connId.String(),
+		)
+	}
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/chainsync.go]: Addressed in 97cf2edf: when the fork-resolution header-queue overflow restart itself fails, the recovery path now clears the queue and forces a resync instead of only logging the failure, so we don't leave blockfetch stuck undrained.

File: ledger/chainsync.go [hunk 6]
```diff
@@ -3114,6 +3186,53 @@ func (ls *LedgerState) restartQueuedBlockfetchAfterForkLocked(
...
+		ls.chain.HeaderCount() == 0 {
+		return
+	}
+	ls.chainsyncBlockfetchMutex.Lock()
+	defer ls.chainsyncBlockfetchMutex.Unlock()
+	if ls.chainsyncBlockfetchReadyChan != nil ||
+		ls.blockfetchContinuationPending {
+		return
+	}
+	if err := ls.startQueuedBlockfetchLocked(connId, pending); err != nil {
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/chainsync.go]: Addressed in 97cf2edf, same commit as the recover-not-just-log fix above: blockfetch peer selection on retarget now re-resolves against the current peer rather than a stale one left over from before the overflow/restart.

File: internal/test/testutil/validated_conway_block.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+) ValidatedConwayBlock {
+	t.Helper()
+
+	vrfPk, vrfSk, err := vrf.KeyGen(seed[:])
+	require.NoError(t, err)
+
+	kesSeed := seed
+	kesSeed[0] ^= 0xAA
+	kesSk, kesPk, err := kes.KeyGen(kes.CardanoKesDepth, kesSeed[:])
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply internal/test/testutil/validated_conway_block.go:99]: Addressed in 97cf2edf: the test helper's KES signing now evolves the signing key to the correct period for the target slot instead of always signing at period 0, so tests exercising non-zero KES periods actually validate the signature they intend to.

File: internal/test/testutil/validated_conway_block.go [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+		headerBodyCbor, encErr := cbor.Encode(headerBody)
+		if encErr != nil {
+			continue
+		}
+		// Store the CBOR on the header body so VerifyBlock's
+		// extractOriginalBodyCbor can retrieve it for KES verification.
+		headerBody.SetCbor(headerBodyCbor)
+
+		kesSig, signErr := kes.Sign(kesSk, 0, headerBodyCbor)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply internal/test/testutil/validated_conway_block.go]: Same KES-period-0 test-helper fix as the coderabbitai comment above, addressed in 97cf2edf.

File: dingo.yaml.example [hunk 9]
```diff
@@ -431,6 +431,20 @@ validateHistorical: true
...
 blockPipelineEnabled: false
 
+# Also VRF/KES-validate blocks in the block-pipeline replay loop with a
+# parallel worker pool (default: false). Requires blockPipelineEnabled to
+# also be set. This adds a second, parallel VRF/KES check immediately before
+# ledger apply; it does not change the existing serial VRF/KES checks that
+# already gate what is admitted to the primary chain, so with
+# validateHistorical: true (the default) blocks are checked twice
+# (redundant but harmless). See ARCHITECTURE.md ("Block Processing
+# Pipeline").
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply dingo.yaml.example]: Addressed in 97cf2edf: reworded the dingo.yaml.example comment -- it no longer claims serial VRF/KES checks remain unchanged/"checked twice" when pipeline validation is enabled; it now describes the actual admission-time skip of the redundant serial re-check.

File: dingo.yaml.example [hunk 10]
```diff
@@ -431,6 +431,20 @@ validateHistorical: true
 # CLI: --block-pipeline-enabled
 blockPipelineEnabled: false
 
+# Also VRF/KES-validate blocks in the block-pipeline replay loop with a
+# parallel worker pool (default: false). Requires blockPipelineEnabled to
+# also be set. This adds a second, parallel VRF/KES check immediately before
+# ledger apply; it does not change the existing serial VRF/KES checks that
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply dingo.yaml.example]: Same stale wording fix in dingo.yaml.example as the coderabbitai comment above, addressed in 97cf2edf.

File: internal/config/config.go [hunk 11]
```diff
@@ -494,43 +494,52 @@ type Config struct {
...
+	BlockPipelineEnabled bool `yaml:"blockPipelineEnabled"         envconfig:"DINGO_BLOCK_PIPELINE_ENABLED"`
+	// BlockPipelineValidateEnabled turns on the block pipeline's VRF/KES
+	// validate stage (issue #1894 phase 3), in addition to decode. Has no
+	// effect unless BlockPipelineEnabled is also set. This adds a second,
+	// parallel VRF/KES check immediately before ledger apply; it does not
+	// change the existing serial VRF/KES checks that already gate what is
+	// admitted to the primary chain, so blocks validated under
+	// ValidateHistorical=true are checked twice (redundant but harmless).
+	// See ARCHITECTURE.md ("Block Processing Pipeline").
+	BlockPipelineValidateEnabled bool `yaml:"blockPipelineValidateEnabled" envconfig:"DINGO_BLOCK_PIPELINE_VALIDATE_ENABLED"`
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply internal/config/config.go:553]: Addressed in 97cf2edf: reworded the internal/config/config.go doc comment to match -- it no longer describes chainsync admission and blockfetch as re-running VRF/KES checks the pipeline already performed; it now documents the state-only check that remains.

File: ledger/block_pipeline_errors_drain_test.go [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+	testutil.RequireReceive(
+		t,
+		done,
+		20*time.Second,
+		"decodeReadChainBatch deadlocked submitting validation-failing "+
+			"blocks whose errors were never drained from errorsChan "+
+			"(issue #1894 regression)",
+	)
+
+	require.Positive(
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/block_pipeline_errors_drain_test.go]: Addressed in 97cf2edf: replaced the racy require.Positive check with require.Eventually (testutil-style polling) so the assertion no longer depends on winning a timing race against the pipeline's asynchronous drain.

File: ledger/chainsync_fork_queue_full_test.go [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+	resolved, err := fixture.ls.tryResolveFork(evt, notFitErr, nil)
+	require.NoError(t, err)
+	require.False(t, resolved)
+
+	assert.Zero(
+		t,
+		requestCount,
+		"an already in-progress blockfetch batch must not be "+
+			"interrupted/restarted by a queue-full fork-extension failure",
+	)
```
[the remark this answers is not in the document]
Comment (cryptodj413) [reply ledger/chainsync_fork_queue_full_test.go:334]: Addressed in 97cf2edf: the test now asserts channel identity (that tryResolveFork kept the same active batch channel) rather than only checking requestCount == 0, which is a weaker signal that doesn't actually prove the existing batch stayed active.

File: .github/workflows/publish.yml [hunk 14]
```diff
@@ -133,11 +133,6 @@ jobs:
   go-test-other:
     name: go-test (${{ matrix.platform }}) (${{ matrix.go-version }})
     needs: [ci, create-draft-release]
```
Comment (wolf31o2) [inline .github/workflows/publish.yml:135]: **Blocking:** the `permissions: contents: read` block that #3226 added to `go-test-other` is gone, so this job goes back to inheriting repository/organization defaults.

That is a least-privilege regression on a release-path job, and it is the only job in this workflow without an explicit grant again — which is precisely what #3226's comment said it was fixing.

File: .github/workflows/publish.yml [hunk 15]
```diff
@@ -175,11 +162,7 @@ jobs:
       - name: go-build
         run: go build ./cmd/dingo
       - name: go-test
-        # Windows runners are substantially slower for this suite. Keep the
-        # same link-size optimization and explicit package timeout used by the
-        # Linux release validation job, and cap the job so a stuck test cannot
-        # hold the release indefinitely.
-        run: go test -ldflags="-s -w" -timeout 20m ./...
+        run: go test ./...
```
Comment (wolf31o2) [inline .github/workflows/publish.yml]: **Blocking:** this drops `-ldflags="-s -w" -timeout 20m` from the test command and `timeout-minutes: 90` from the job, both added by #3226.

The `-timeout 20m` matters most, and it matters most *for this PR*. Dingo's own guidance is that `ledger` runs roughly 9-13 minutes under `-race`, past `go test`'s 10-minute default — and this PR adds 52 tests to that package, several of them pipeline-lifecycle and goroutine-leak tests. Reverting to the default timeout makes a `panic: test timed out` on the Windows release job more likely, not less, and it names whichever test held the alarm rather than the cause.

Losing `timeout-minutes: 90` also removes hang protection: the job falls back to the 360-minute default, so a genuinely stuck test can hold `build-binaries` and `finalize-release` for six hours.

File: ledger/state.go [hunk 16]
```diff
@@ -3341,6 +3705,68 @@ func (ls *LedgerState) ledgerReadChainIterator(
...
+// errors when it reads the corresponding item back from Results(), but that
+// only covers items that make it that far -- this is the only path that
+// also covers, e.g., apply-stage invariant violations such as
+// pipeline.ErrBlockNotValidated) and is logged at error level plus its own
+// counter for operator visibility.
+func (ls *LedgerState) recordBlockPipelineError(err error) {
+	if err == nil {
+		return
+	}
+	if errors.Is(err, errBlockPipelineEta0Unavailable) {
```
Comment (wolf31o2) [inline ledger/state.go]: **Non-blocking (observability).** This classification keys purely on the `errBlockPipelineEta0Unavailable` sentinel, with no era check — unlike `decodeReadChainBatch`, which correctly gates the equivalent ignore on `block.Era().Id != byron.EraIdByron`. Enforcement is unaffected either way; this is only about what an operator sees. It misreports in both directions:

**Praos-era nonce gap reads as expected.** `headerVerificationEpoch` returns `errEpochNonceUnavailable` whenever `len(epoch.Nonce) == 0`, and its own comment documents that as "the epoch rollover has not completed" — a Praos-era condition, which the serial path treats as a rejection. Here it lands at `Debug` and increments `blockPipelineExpectedEta0Errors`. I did not reproduce it, so I am relying on that comment rather than claiming I hit it.

**Transient cache miss reads as unexpected.** With `allowEpochCacheAdvance=false`, a slot the epoch cache does not yet cover returns `errHeaderVerificationDeferred`, which is *not* wrapped in the sentinel — so it takes the `Error` branch and `blockPipelineUnexpectedErrors`, once per block per retry. `ARCHITECTURE.md` describes exactly that case as normal and recoverable "once the epoch cache catches up".

Net effect is that the one condition where a deterministic rejection stalls sync is the quiet one, while the transient that self-heals is the loud one. Mirroring `decodeReadChainBatch`'s era check, or classifying on deferred-vs-nonce-missing rather than on the single sentinel, would line the counters up with the operational meaning.

`TestRecordBlockPipelineErrorClassification` covers the `%w`-vs-string wrapping mechanics well, but not either case above.

**Review by wolf31o2 — CHANGES_REQUESTED** [review]
The ledger work carries no correctness defect that this review could establish. Changes are requested for one reason: the merge commit dropped three CI hardening measures present on `main`. Inline comments cover those, plus one non-blocking observability point.

**Verified against the code, not the comments**

- `blockPipelineEta0Provider`'s concurrency claim holds. With `allowEpochCacheAdvance=false`, `headerVerificationEpoch` returns before `ensureEpochForSlot`, so the validate workers provably never mutate epoch-cache state.
- The `errorsChan` drain is genuinely required, not defensive. Confirmed in gouroboros v0.193.3 that `StageWorkerPool.worker` pushes every error before forwarding, and that `BlockPipeline.Stop` closes `errorsChan` (`pipeline.go:303`). Single `Start`/`Stop` pair, so the drain goroutine really does live for the pipeline's lifetime, and `Close`'s wait is bounded by `CloseBlockPipelineDrainTimeout`.
- `blockPipelineGatherMutex` discipline is correct. Every early return inside the gather loop calls `releaseGatherLock()`, and it is called unconditionally after `decodeReadChainBatch` before any post-loop return; `gatherLockHeld` re-initialises per outer pass. No path leaks the read lock into a rollback's write lock.
- Enforcement is not weakened by the admission-time skip. `decodeReadChainBatch` gates on `block.Era().Id != byron.EraIdByron` and aborts the whole batch, so the Byron ignore cannot silently pass a Praos-era block.
- `BlockPipelineValidateEnabled` defaults off and requires `BlockPipelineEnabled`, with a config-time error for the invalid combination.

**Checks run locally at 97cf2edf**

| Command | Result |
| --- | --- |
| `go test -race -timeout 25m` (52 new/changed tests, `./ledger/ ./internal/config/`) | pass — ledger 47.7s, config 2.9s |
| `make docs-parity` | pass |
| `make import-boundaries` | pass |

**Devnet and conformance:** both run and passing — results and their caveat are in a follow-up comment. CI's 7 checks are green but cover neither.

`ARCHITECTURE.md` documents the `errorsChan` deadlock, the Byron-only ignore, and that headers queue via `AddBlockHeader` (unverified) when the pipeline revalidates. No documentation gap found.

File: .github/workflows/publish.yml (hunk 14 above)
Comment (cryptodj413) [reply .github/workflows/publish.yml:135]: Addressed in fd51dff1: restored `go-test-other`'s `permissions: contents: read` grant, `timeout-minutes: 90`, and `go test -ldflags="-s -w" -timeout 20m ./...` verbatim from main. The merge-conflict resolution had picked the pre-#3226 version of this job; `publish.yml` now diffs clean against `origin/main`.

File: .github/workflows/publish.yml (hunk 15 above)
Comment (cryptodj413) [reply .github/workflows/publish.yml]: Same fix as the permissions-grant comment above, addressed in the same commit (fd51dff1): restored -ldflags="-s -w" -timeout 20m on the go-test command and timeout-minutes: 90 on the job, verbatim from main.

File: ledger/state.go (hunk 16 above)
Comment (cryptodj413) [reply ledger/state.go]: Addressed in fd51dff1: recordBlockPipelineError now distinguishes errHeaderVerificationDeferred (transient epoch-cache lag behind an already-committed block, per ARCHITECTURE.md self-healing once the cache catches up) from the eta0-unavailable case, with its own dingo_ledger_block_pipeline_deferred_epoch_cache_errors_total counter and debug log instead of falling into the unexpected-error bucket. I did not add an era check here, since errorsChan carries a bare error with no item/era context to check against -- only decodeReadChainBatch has the decoded block, and it already gates enforcement on block.Era().Id independently, so enforcement is unaffected either way. I did soften the eta0-unavailable debug log to stop claiming the condition is Byron-specific, since that isn't verified at this point either. New regression test: TestRecordBlockPipelineErrorClassificationDeferredIsNotUnexpected.

**Comment by cryptodj413** [comment]
Addressed all three review comments (fd51dff1):
- Restored the `go-test-other` CI hardening (`permissions: contents: read`, `timeout-minutes: 90`, `-ldflags="-s -w" -timeout 20m`) dropped by the merge-conflict resolution; `publish.yml` now diffs clean against `origin/main`.
- Split `errHeaderVerificationDeferred` into its own expected/transient counter (`dingo_ledger_block_pipeline_deferred_epoch_cache_errors_total`) in `recordBlockPipelineError`, instead of counting it as unexpected; softened the eta0-unavailable debug log to not claim Byron-specificity it can't verify. New test: `TestRecordBlockPipelineErrorClassificationDeferredIsNotUnexpected`.

Checks run locally at fd51dff1: `go test -race -timeout 25m ./ledger/... ./internal/config/...` (pass, ledger 536.8s), `make docs-parity`, `make import-boundaries`, `make gorm-check`, `golangci-lint run`, `nilaway`, `modernize` -- all clean. CI (7 checks) and cubic's fresh pass (0 issues) are green.

`ARCHITECTURE.md` updated to match the new classification.

@wolf31o2 ready for another look.

**Comment by wolf31o2** [comment]
The devnet and conformance suites both pass at `97cf2edf`.

| Suite | Command | Result |
| --- | --- | --- |
| Ledger conformance vectors | `go test ./internal/test/conformance/` | pass, 6.4s |
| DevNet, accelerated timeline | `internal/test/devnet/run-tests.sh --accelerated` | pass — `TestAcceleratedScenarioTimeline` 120.89s |

DevNet detail: all four nodes agreed on the same chain hash at slot 192 across the relay-restart scenario, the relay recovered to tip slot 198 / block 45, and txpump reported `submitted=3845 rejected=27 error=0`.

**Both suites ran with `BlockPipelineValidateEnabled` off**, because it defaults off and `internal/test/devnet/` contains no reference to the flag:

```
$ grep -rn "BLOCK_PIPELINE\|block-pipeline\|blockPipeline" internal/test/devnet/
(no matches)
```

These runs therefore establish only that the change does not regress the existing serial path. The validate stage itself has unit coverage but no end-to-end coverage in any suite. Since that stage skips admission-time crypto and queues headers via `AddBlockHeader` unverified, an accelerated devnet run with the flag enabled is owed before it is switched on anywhere real; the deterministic-rejection stall path has no integration exercise at all.

Not run: `--conformance` mode (dingo + `cardano-node` reference topology), which is the relevant one for a header-crypto change.

Unrelated and unchanged: the merge in `97cf2edf` dropped CI hardening present on `main`.

**Comment by cryptodj413** [comment]
Thanks for running both suites and for catching your own "not run" claim -- good to have the actual devnet/conformance pass on record at 97cf2edf.

Agreed on the coverage gap: BlockPipelineValidateEnabled has unit coverage but nothing end-to-end, and wiring it into internal/test/devnet/ is a real (if small) infra change, not something to fold into this already-large PR. Opened #3242 to track adding the env passthrough and an accelerated run with the flag on, rather than deferring it silently -- same pattern as #3227 for the phase 2 apply-stage gap.

Re the blocking finding: that's fixed as of fd51dff1 (see my reply on that thread) -- restored publish.yml's permissions/timeout-minutes/-ldflags+-timeout verbatim from main. Let me know if you want the --conformance run too; happy to kick it off if useful alongside #3242.

**Review by wolf31o2 — APPROVED** [review]
